File: 67.add-binary.java
/*
 * @lc app=leetcode id=67 lang=java
 *
 * [67] Add Binary
 */

// @lc code=start
class Solution {
    public String addBinary(String a, String b) {
        if(a == null || a.isEmpty()) {//Kiểm tra chuỗi null hoặc rỗng
            return "0";//Khi chuỗi null hoặc rỗng thì trả về "0"
        }
        if(b == null || b.isEmpty()) {//Kiểm tra chuỗi null hoặc rỗng
            return "0";//Khi chuỗi null hoặc rỗng thì trả về "0"
        }
        StringBuilder result = new StringBuilder();//Biến để lưu kết quả
        int carry = 0;//Biến để lưu giá trị nhớ
        int i = a.length() - 1;//Chỉ số bắt đầu từ cuối chuỗi a
        int j = b.length() - 1;//Chỉ số bắt đầu từ cuối chuỗi b
        //Bắt đầu cộng từ cuối chuỗi a và b
        while(i >= 0 || j >= 0 || carry > 0) {//Kiểm tra điều kiện dừng
            if(i >= 0) {//Kiểm tra chỉ số i có hợp lệ không
                carry += a.charAt(i--) - '0';//Lấy giá trị của ký tự tại chỉ số i và cộng vào carry
            }
        if(j >= 0) {//Kiểm tra chỉ số j có hợp lệ không
            carry += b.charAt(j--) - '0';//Lấy giá trị của ký tự tại chỉ số j và cộng vào carry
        }
        result.append(carry % 2);//Thêm giá trị của carry vào kết quả
        carry /= 2;//Cập nhật giá trị của carry
    }
    return result.reverse().toString();//Trả về kết quả sau khi đảo ngược
}
}
// @lc code=end

File: 67.add-binary.py
#
# @lc app=leetcode id=67 lang=python3
#
# [67] Add Binary
#

# @lc code=start
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        if not a or not b: # Nếu một trong hai chuỗi rỗng
            return a or b # Nếu cả hai chuỗi đều rỗng, trả về chuỗi không rỗng
        max_len = max(len(a), len(b)) # Tính chiều dài tối đa
        a = a.zfill(max_len) # Điền thêm số 0 vào đầu chuỗi a
        b = b.zfill(max_len) # Điền thêm số 0 vào đầu chuỗi b
        carry = 0 # Biến nhớ để lưu trữ giá trị nhớ
        result = [] # Danh sách để lưu trữ kết quả
        for i in range(max_len - 1, -1, -1): # Duyệt từ phải sang trái
            total = carry + (1 if a[i] == '1' else 0) + (1 if b[i] == '1' else 0) # Tính tổng
            result.append('1' if total % 2 == 1 else '0') # Thêm bit kết quả vào danh sách
            carry = 1 if total > 1 else 0 # Cập nhật giá trị nhớ
        if carry: # Nếu còn giá trị nhớ
            result.append('1') # Thêm bit nhớ vào danh sách
        return ''.join(reversed(result)) # Trả về chuỗi nhị phân kết quả
# @lc code=end

